### ISSUE TITLE
Fold does not get decorators for plugin

### ISSUE THREAD

**Issue by OskarLebuda** [swc-project/swc#9024]
### Describe the bug

We are working on moving into Rspack . To make that happen we have to rewrite our custom TypeScript transformer (AST) into SWC plugin. The plugin for reproductions is available [here](https://github.com/enetproduction/swc-plugin-vue-auto-prop)

#### What is the problem?
The problem is that class member has as empty array of decorators even if they exists (see example)
In the SWC Playground decorators also exists in as an filled array.  
But inside of our code the decorators array is empty:
![image](https://github.com/swc-project/swc/assets/10373127/bb003f9f-34b8-4f29-ac27-62225d9c8d5d)
```
The prop of member class is ClassProp(ClassProp {
    span: 84..118#0,
    key: Ident(Ident {
        span: 104..108#0,
        sym: "test", optional: false
    }),
    value: None,
    type_ann: None,
    is_static: false,
    ==========================
    decorators: [],
    ==========================
    accessibility: None,
    is_abstract: false,
    is_optional: false,
    is_override: false,
    readonly: false,
    declare: false,
    definite: false
})
```

The problem is that [all tests](https://github.com/enetproduction/swc-plugin-vue-auto-prop/blob/main/transform/src/lib.rs#L71) are working properly and transforms class property decorator into new value but the example code does not. 

#### How to reproduce?
We've been created an reproduction [repository](https://github.com/enetproduction/swc-plugin-vue-auto-prop). <br>
The example code is available into `example` directory. To run the test you have to install all required packages and run `build` command:
```
$ pnpm install
$ cd example && pnpm install
$ pnpm build
```

#### What is the expected output?
We are trying to make plugin that will add into `Prop` decorator the annotation type e.g.
Before:
```typescript
export default class Test {
  
  @Prop()
  public firstProp: boolean;

  @Prop({ default: boolean })
  public secondProp: boolean;

  @Prop()
  public thirdProp: string;

}
```

After:
```typescript
export default class Test {
  
  @Prop({ type: Boolean })
  public firstProp: boolean;

  @Prop({ default: boolean, type: Boolean })
  public secondProp: boolean;

  @Prop({ type: String })
  public thirdProp: string;

}
```

### Input code

```typescript
import { Prop } from "./decorators/Prop";

export default class TestComponent {

    @Prop()
    private test: boolean;

}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false,
      "decorators": true
    },
    "target": "es2019",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    },
    "experimental": {
        "plugins": [
            ["@enp/swc-plugin-vue-auto-prop", {}],
        ]
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```
```


### Playground link (or link to the minimal reproduction)

https://play.swc.rs/?version=1.5.24&code=H4sIAAAAAAAAAx2MMQqEQAxF%2B5wiWGmz9toIXsBiL5DVCAPjZMhklwWZuxv91Yf%2F%2FgtHFjU8F5VccVc5sHn1G6%2BiZKKlv4dmBOD%2FA2680zcarpFKwTcXm8UViZNLANAz3Ze2e3rW8CNjNAcH%2FIhEpuS2ChePRyj4eQAAAA%3D%3D&config=H4sIAAAAAAAAA2WPSw7CMAxE9z1F5TULYIEEd%2BAQUepWQc1HtisRVb17kzYNIHbxvHjGMzdtCy%2FW8Gjn9ExDUMRIdU4KRyfqnRSQGJA1mSBwOqhwRr0aGavWofakxBMnJDThBpadgygaULIf8vV8uRcvGL1n%2FPUCa5zp4%2Fc12ttAyPwXapUbxrq%2FJzYlFazvpg2WlrnJfsENPp%2BOsGoMhp%2FH5tZjWQEa0ouIMQEAAA%3D%3D

### SWC Info output

```
    Operating System:
        Platform: darwin
        Arch: arm64
        Machine Type: arm64
        Version: Darwin Kernel Version 23.5.0: Wed May  1 20:14:38 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6020
        CPU: (10 cores)
            Models: Apple M2 Pro

    Binaries:
        Node: 16.20.2
        npm: 8.19.4
        Yarn: 1.22.21
        pnpm: 9.1.3

    Relevant Packages:
        @swc/core: 1.5.25
        @swc/helpers: N/A
        @swc/types: 0.1.7

    SWC Config:
        output: N/A
        .swcrc path: N/A

    Next.js info:
        output: N/A
```

### Expected behavior

The `decorators` arrays are filled.

### Actual behavior

_No response_

### Version

latest

### Additional context

_No response_

**Comment by kwonoj** [comment]
If I remember correctly, this is currently expected behavior. There are few pre-transform passes before running plugin transform such as typescript strip and decorators for the performance & few other reasons.

**Comment by kwonoj** [comment]
Not going to say expected behavior === will do nothing, but also we had some prior discussions regarding those prepass transforms and found it wasn't super ideal to pass everything into plugin.

**Comment by OskarLebuda** [comment]
> Not going to say expected behavior === will do nothing, but also we had some prior discussions regarding those prepass transforms and found it wasn't super ideal to pass everything into plugin.

Right. The question is what should I do? Write custom transform? Stripe? 
Is there any kind of docs how should I make it happen?

Without this transformation we cannot use Rspack and swc

**Comment by kwonoj** [comment]
Unfortunate to say as of today, the only way I can think of is having custom transformation and create a swc binary instead of using prepublished one. As you said, that'll defeat to use any other packages like rspack.

Other than those, I do not have an immediate suggestion. Adding additional transform phase in plugin transform might need some proof if it won't hurt any other generic plugin transforms.

**Comment by OskarLebuda** [comment]
That’s bad information for us :( 
Do you consider to create some hooks for compilation?
Or maybe even JS API to make this work?

**Comment by kwonoj** [comment]
That's more no-go given we skipped some transform for the performance reasons.

**Comment by OskarLebuda** [comment]
This could be nice feature for SWC. There are tones of plugins for babel typescript that can be rewrites as SWC transformer. 

Do you have some docs how should we build and use custom binary of SWC?

**Comment by kwonoj** [comment]
> build and use custom binary of SWC

You may check how swc builds itself.

https://github.com/swc-project/swc/tree/main/bindings/binding_core_node

> This could be nice feature for SWC. There are tones of plugins for babel typescript that can be rewrites as SWC transformer.

I do not want derail this thread for those topics, there are huge lengthy discussions around plugin architectures and there's no way 1:1 babel - swc transformer conversion anyway.

I'm moving this to discussion, since this is not an actual bug but more of feature request now.
